Source/CarthageKit/NewResolver.swift

```diff
@@ -0,0 +1,555 @@
+	var edges: [DependencyNode: Set<DependencyNode>] = [:]
+
+	/// List of nodes that still need to be processed, in the order they should be processed
+	var unvisitedNodes: [DependencyNode] = []
+
+	/// Whitelist of dependencies to limit updates to. Once a valid graph is found,
+	/// the nodes are filtered down to this list. Remaining nodes are given the values found in 'lastResolved'.
+	/// The graph is then re-checked for validity. If it's now invalid due to the whitelist, the search is continued
+	let whitelist: [String]?
+	let lastResolved: [Dependency: PinnedVersion]?
+
+	init(whitelist: [String]?, lastResolved: [Dependency: PinnedVersion]? = nil) {
+		self.whitelist = whitelist
+		self.lastResolved = lastResolved
+	}
+
+	/// A dictionary defining all the versions that are pinned in this graph.
+	var versions: [Dependency: PinnedVersion] {
+		var versionDictionary: [Dependency: PinnedVersion] = [:]
+		for node in allNodes {
+			versionDictionary[node.dependency] = node.proposedVersion
+		}
+		return versionDictionary
+	}
+
+	/// Returns the next unvisited node if available
+	mutating func nextNodeToVisit() -> DependencyNode? {
+		return !unvisitedNodes.isEmpty ? unvisitedNodes.removeFirst() : nil
+	}
+
+	/// Runs final validation against the whitelist on a complete graph (e.g., no more unvisited nodes)
+	///
+	/// For every node that is not in the whitelist or its dependencies
+	/// - If it has a version in 'lastResolved', it must match, or an unsatisfiableDependencyList error is returned
+	/// - If the node has no previous version, it will be removed from the returned graph, but the graph is still considered valid
+	func validateFinalGraph() -> Result<DependencyGraph, CarthageError> {
+		guard unvisitedNodes.isEmpty else {
+			return .failure(.internalError(description: "Validating graph before it's been completely expanded"))
+		}
+
+		guard let whitelist = whitelist, !whitelist.isEmpty, let lastResolved = lastResolved else {
+			return .success(self)
+		}
+
+		// Any dependencies of items in the whitelist are also allowed to update
+		var nodeWhitelist = Set<DependencyNode>()
+		allNodes
+			.filter { whitelist.contains($0.dependency.name) }
+			.forEach { node in
+				nodeWhitelist.insert(node)
+				if let nestedDependencies = edges[node] {
+					nodeWhitelist.formUnion(nestedDependencies)
+				}
+			}
+
+		var filteredGraph = self
+		for node in allNodes {
+			guard !nodeWhitelist.contains(node) else {
+				continue
+			}
+
+			guard let lastVersion = lastResolved[node.dependency] else {
+				// If it doesn't have a previous version, and isn't in the whitelist, remove it from the returned graph
+				filteredGraph.allNodes.remove(node)
+				filteredGraph.edges.removeValue(forKey: node)
+				continue
+			}
+
+			// If it's not in the whitelist, and it has a previous version, they should match
+			if lastVersion != node.proposedVersion {
+				return .failure(.unsatisfiableDependencyList(whitelist))
+			}
```

Comment:
This check seems to be new in this resolver and I don't quite understand whether (a) this failure would indicate a logic error in Carthage or (b) an incompatibility based on the constraints. Can you elaborate about when this would occur?

Reply:
This actually fixes a bug that exists in the current resolver. It's the formerly 'pending' test (which is now wrapped in a resolver type check). (See the changes below in the test class)

The previous resolver would:
- Find a valid graph
- filter out any non-whitelisted nodes from the graph, assigning them back to their 'lastResolved' value

This actually had the potential to result in an invalid graph, that was never subsequently validated (that's actually what the pending test is for). Because the 'lastResolved' may be incompatible with the new values for the nodes that were updated.

Instead, the new resolver has a slight variant, it now:
- Find a valid graph
- filter out any non-whitelisted nodes from the graph, and *validates that their values == the 'lastResolved' value*

Instead of attempting to figure out what the 'right' graph is post-filter, I'm just relying on the permutations to eventually find a graph that only updates the whitelisted nodes.

Reply:
This seems like too strict a check. We should be validating that the nodes that aren't updated are _compatible_ with the version specifiers—not that they resolve to the same versions.

Reply:
That would make for a good additional test.

If one of the dependencies not in the whitelist has a new minor version, updating the whitelist should still succeed.

Reply:
I'm not sure I get what you're saying. The point of the whitelist is that anything _not_ in it doesn't change at all. That's what this check is doing. It's saying: "If you're not in the whitelist, you should be what you were before"

Reply:
It seems like either:

1. `node.proposedVersion` can only equal `lastVersion` because it's set to that explicitly during resolution. In this case, the check is redundant and could be eliminated.

2. `node.proposedVersion` is a result of normal resolution, in which case it could be e.g. `3.2.4` when `lastVersion = 3.2.3`. If the version specifier for this dependency is `~> 3.2`, then  we're making the whitelist much less useful unnecessarily.

Reply:
if we're allowing things outside the whitelist to be updated if they are still compatible with their version specifiers... isn't that not a whitelist? Isn't that functionally equivalent to just running the resolver on everything?

Reply:
I'm not saying update them. I'm saying we should allow the update of the whitelist to complete if one of the items outside the whitelist doesn't exactly match the last resolved version—so long as the last resolved version is compatible.

Reply:
This is obviously a bit confusing. If it'd help, I can put together a test case myself and submit it as a separate PR.

Reply:
I think I get it, it's an optimization not a correctness thing. If we come across a permutation that works within the specifiers, we could adjust the graph without breaking the validity of it. 

Currently it would reject it, but eventually 'come around' to the same graph when the permutations catch up to the same result, with matching proposed versions. 

I'll look into it, however that may be a problem due to how the resolution occurs in the first place. I don't "merge" version specifiers anymore when coming across children, I just reject outright if the children version specifier excludes the current permutation. 

What this means is at the point where I'm validating the whitelist, it may appear that the 'lastResolved' is compatible with the version specifier, but in fact it had been rejected by a child